language-support/ts/daml-ledger/index.ts

```diff
@@ -502,7 +502,7 @@ class Ledger {
     const protocols = ['jwt.token.' + this.token, 'daml.ws.auth'];
     let ws = new WebSocket(this.wsBaseUrl + endpoint, protocols);
     let isLiveSince: undefined | number = undefined;
-    let lastOffset: undefined | string = undefined;
+    let lastOffset: undefined | null | string = undefined;
     let state = init;
     let isReconnecting: boolean = false;
     const emitter = new EventEmitter();
```

Comment:
You need to change how we handle reconnects. The JSON API can return a `null` offset when it was ledger begin but it only accepts strings. The only way to get a null offset is if there were literally no transactions, so we basically reset and treat this as the initial connect with no event. If you want to be safe, reset `state` to `init` although afaict this isn’t necessary since the list of events will have been empty so we never called `change`.

Reply:
Sorry not up to date on this issue. Is that considered correct behaviour from the JSON API? It seems really weird it would produce values it cannot consume, especially for a parameter that is specifically meant as "this is an opaque thing you need to give back to me later on".

Reply:
It’s at least documented behavior. It seems reasonable to allow this but I would consider this a separate issue. Given that we can fix this only in the typescript libraries, I would fix it here for now.

There is also a bug in the JSON API where it returns `null` when it shouldn’t but that’s a separate issue. The issue addressed here cannot be hit on Sandbox since it will never return ledger begin on the ACS (but other ledgers might).

Reply:
So now we don't send the `offset` message when our last offset is `null`. My understanding is that this will start the stream from the very beginning of transactions which is correct here.